etc/docs/Design.md

```diff
@@ -0,0 +1,125 @@
+
+### Streaming data
+
+In order to be able to stream and also send the full output of the processes we need to store the output of this in memory for simplicty. See [Tradeoffs](#tradeoffs) section for information about why this solution should not be used in production  
+
+## Security
+
+### Cipher suite
+
+In order to have the latest and strongest transport encryption TLS 1.3 should be utilized. The cipher suite TLS_AES_128_GCM_SHA256 is the mandatory cipher suite for TLS 1.3, and it provides sufficiently strong encryption there for it will be used by the server
+The decision to ultimately use TLS 1.3 is due to the fact that the client and the server controlled by the same entity, therefore it's easy to enforce the usage of the latest version
+
+### Authentication
+
+### mTLS
+
+For the mTLS the following certificates will be generated by a generate-certs shell script to enable easy testing of the application (see the Tradeoff section for more details):
+
+* A root CA certificate to sign client and server CSR-s 
+* A x509 certificate for the server signed by the root CA
+* Two x509 certificates one for each of two clients, that could be used to connect to the applicaiton
+
+The root CA will use a private key that's 4096 bits and the client and server certs will use private keys with 2048 bits
+
+### Authorization
+
+To keep things simple, the following statement should hold for every process started on the server:
+
+`Only clients authenticated with the client certs having the same CN can get any information about, or interact with a certain process`
+
+## Process execution
+
+### Resource control
+
+To control resource allocation linux control groups (chroups) can be utilized
+
+For the required CPU, Memory, and DiskIO limits a new cgroup could be created per process using the cpu, memory and blkio controllers
+
+### Isolation
+
+To implement isolation of namespaces for PID, Mount and Networks unix namespaces will be used
+
+In order to be able to set up proper isolation there is need to run some initialization process before the app executes in the new namespace.
+
+## Client
+
+The client CLI will have a two top level commands:
+
+`start` and `execution`
+
+The `start` command will send a command to be started to the server, along with the arguments passed to it
+
+```
+Usage: args start [--command COMMAND] [ARGS [ARGS ...]]
+
+Positional arguments:
+  ARGS                   these arguments will be supplied to the executed command
+
+Options:
+  --command COMMAND      the command to be executed on the server
+
+```
+
+The example of `cu-thor start --command head /dev/random` will result in executing the command `head` with the argument `/dev/random`
+Similarly the command `cu-thor start --command head -- -n10 /dev/random` will execute the command `head` with the arguments `-n10` and `/dev/random`
+
+The `execution` will have a `list` subcommand that will return the current executions that this client has rights to interact with
```

Comment:
My original thinking was that `execution` commands always require an execution id, this is a bit broken by the `list` subcommand, which then should probably be a top-level command to uphold this invariant. 

If I were to move start under execution, then `execution` itself would not make sense because every command is an execution.

So the logic is the following `start` -> there is no execution to interract with, `execution` it interacts with a previously `start`-ed execution

Reply:
gotcha
this may be one of those UX decisions that makes more sense for the creator (you) but is confusing for the user that has minimal understanding of the system

Reply:
fair enough, as at this point you are the closest I have to a user, I will treat this conversation as user research and adapt the UX accordingly